### PR TITLE
adapter: require Session's external metadata rx at construction

### PR DESCRIPTION
**Pull request by benesch**
Easy split out from https://github.com/MaterializeInc/materialize/pull/25409.

Rather than allowing registration of an external metadata receiver at any point after a session's construction, require it as part of the construction of the session. This allows us to statically remove the possibility of double registration.

### Motivation

   * This PR refactors existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: src/environmentd/src/http.rs
Comment (benesch) [inline src/environmentd/src/http.rs:882]: @mjibson I added a comment to the code based on your question [here](https://github.com/MaterializeInc/materialize/pull/25409/files#r1498425287).

File: src/sql/src/session/vars.rs
Comment (benesch) [inline src/sql/src/session/vars.rs:2739]: @mjibson noted this was a behavior change [here](https://github.com/MaterializeInc/materialize/pull/25409/files#r1498097598).

Definitely worth discussing further. I don't understand why the current code exempts external admins from the connection limits. The max connections limit is to protect _us_, right? ISTM that we should not be allowing any end users to slip past the connection limit, even if they're an "Organization Admin".

Comment (mjibson) [reply src/sql/src/session/vars.rs:2739]: I believe the whole point of allowing external admins was so that they could come in and close the connections that were causing the limit to be reached (if say some long running connections got stuck for some reason in their infra). Just seems kinda bad to make the org admins sometimes not able to log in because non-admins are logged in? That was our reasoning at the time.

Comment (ParkMyCar) [reply src/sql/src/session/vars.rs:2739]: I'm leaning towards retaining the original behavior, but definitely should add a comment because I didn't know the reasoning either!

**Review by ParkMyCar — APPROVED** [review]
Plumbing of the `external_metadata_rx` looks good to me, much cleaner!

File: src/sql/src/session/vars.rs
Comment (benesch) [reply src/sql/src/session/vars.rs]: Ok, I figured out how to preserve the existing behavior. Sorry for the noise! I'll file an issue to follow up on the `max_connections` issue.

**Comment by benesch** [comment]
@mjibson should I wait for your approval as well?

**Review by mjibson — APPROVED** [review]

**Pull request merged**
